src/shims/sync.rs

```diff
@@ -326,34 +413,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         let kind = mutex_get_kind(this, mutex_op)?.not_undef()?;
         let locked_count = mutex_get_locked_count(this, mutex_op)?.to_u32()?;
-
-        if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? {
-            if locked_count != 0 {
-                mutex_set_locked_count(this, mutex_op, Scalar::from_u32(0))?;
-                Ok(0)
+        let owner_thread: ThreadId = mutex_get_owner(this, mutex_op)?.to_u32()?.into();
+
+        if owner_thread != this.get_active_thread()? {
+            throw_ub_format!("called pthread_mutex_unlock on a mutex owned by another thread");
+        } else if locked_count == 1 {
+            let blockset = mutex_get_or_create_blockset(this, mutex_op)?;
+            if let Some(new_owner) = this.unblock_some_thread(blockset)? {
+                // We have at least one thread waiting on this mutex. Transfer
+                // ownership to it.
+                mutex_set_owner(this, mutex_op, new_owner.to_u32_scalar())?;
             } else {
-                throw_ub_format!("unlocked a PTHREAD_MUTEX_NORMAL mutex that was not locked");
-            }
-        } else if kind == this.eval_libc("PTHREAD_MUTEX_ERRORCHECK")? {
-            if locked_count != 0 {
+                // No thread is waiting on this mutex.
                 mutex_set_locked_count(this, mutex_op, Scalar::from_u32(0))?;
```

Comment:
We do not reset the owner here. Should we?
If not, maybe adjust the comment above that explains the fields to something like "last thread who acquire this lock" or so, to avoid the "when count > 0" conditional.